File: nice-traffic-light/r/R/classifier.R
# Parse the Evagene /risk/calculate response into a NICE outcome.
#
# Strict by design -- a silent default would mask a server-side schema
# change and leave the caller reasoning over stale assumptions.

known_categories <- c("near_population", "moderate", "high")

#' Raise a response_schema_error condition.
#' @noRd
stop_schema <- function(message) {
  stop(structure(
    list(message = message, call = sys.call(-1)),
    class = c("response_schema_error", "error", "condition")
  ))
}

#' Parse an Evagene risk-calculation response into a nice_outcome.
#'
#' @param payload A named list (e.g. from `jsonlite::fromJSON(..., simplifyVector = FALSE)`).
#' @return A `nice_outcome` list with fields: counselee_name, category
#'   (one of "near_population", "moderate", "high"),
#'   refer_for_genetics_assessment, triggers, notes.
classify_nice_response <- function(payload) {
  require_object(payload, "response")
  cancer_risk <- require_object_field(payload, "cancer_risk")

  outcome <- list(
    counselee_name = optional_string(payload, "counselee_name"),
    category = parse_category(require_string_field(cancer_risk, "nice_category")),
    refer_for_genetics_assessment = require_boolean_field(cancer_risk, "nice_refer_genetics"),
    triggers = require_string_list_field(cancer_risk, "nice_triggers"),
    notes = require_string_list_field(cancer_risk, "notes")
  )
  class(outcome) <- "nice_outcome"
  outcome
}

parse_category <- function(raw) {
  if (!raw %in% known_categories) {
    stop_schema(sprintf("Unknown NICE category: '%s'", raw))
  }
  raw
}

require_object <- function(value, label) {
  if (!is.list(value) || (length(value) > 0 && is.null(names(value)))) {
    stop_schema(sprintf("%s is not an object", label))
  }
  invisible(value)
}

require_object_field <- function(container, key) {
  value <- container[[key]]
  if (!is.list(value) || is.null(names(value))) {
    stop_schema(sprintf("field '%s' is missing or not an object", key))
  }
  value
}

require_string_field <- function(container, key) {
  value <- container[[key]]
  if (!is.character(value) || length(value) != 1L || is.na(value)) {
    stop_schema(sprintf("field '%s' is missing or not a string", key))
  }
  value
}

require_boolean_field <- function(container, key) {
  value <- container[[key]]
  if (!is.logical(value) || length(value) != 1L || is.na(value)) {
    stop_schema(sprintf("field '%s' is missing or not a boolean", key))
  }
  value
}

require_string_list_field <- function(container, key) {
  value <- container[[key]]
  if (is.null(value)) {
    return(character(0))
  }
  if (is.list(value)) {
    # jsonlite(simplifyVector = FALSE) returns a list of length-1 character
    # scalars for a JSON string array.  We must validate each element
    # explicitly -- unlist() would silently coerce mixed types.
    is_string_scalar <- vapply(
      value,
      function(item) is.character(item) && length(item) == 1L && !is.na(item),
      logical(1L)
    )
    if (!all(is_string_scalar)) {
      stop_schema(sprintf("field '%s' is not a list of strings", key))
    }
    return(vapply(value, as.character, character(1L)))
  }
  if (!is.character(value) || any(is.na(value))) {
    stop_schema(sprintf("field '%s' is not a list of strings", key))
  }
  value
}

optional_string <- function(container, key) {
  value <- container[[key]]
  if (!is.character(value) || length(value) != 1L || is.na(value)) {
    return("")
  }
  value
}

File: nice-traffic-light/node/src/classifier.ts
/**
 * Parse the Evagene risk/calculate response into a domain NiceOutcome.
 *
 * Strict by design — a silent default here would mask a server-side
 * breaking change.
 */

export type RiskCategory = 'near_population' | 'moderate' | 'high';

export interface NiceOutcome {
  readonly counseleeName: string;
  readonly category: RiskCategory;
  readonly referForGeneticsAssessment: boolean;
  readonly triggers: readonly string[];
  readonly notes: readonly string[];
}

export class ResponseSchemaError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ResponseSchemaError';
  }
}

const KNOWN_CATEGORIES: readonly RiskCategory[] = ['near_population', 'moderate', 'high'];

export function classifyNiceResponse(payload: unknown): NiceOutcome {
  const root = requireObject(payload, 'response');
  const cancerRisk = requireObjectField(root, 'cancer_risk');
  return {
    counseleeName: optionalString(root, 'counselee_name'),
    category: parseCategory(requireStringField(cancerRisk, 'nice_category')),
    referForGeneticsAssessment: requireBooleanField(cancerRisk, 'nice_refer_genetics'),
    triggers: requireStringListField(cancerRisk, 'nice_triggers'),
    notes: requireStringListField(cancerRisk, 'notes'),
  };
}

function parseCategory(raw: string): RiskCategory {
  const match = KNOWN_CATEGORIES.find(category => category === raw);
  if (match === undefined) {
    throw new ResponseSchemaError(`Unknown NICE category: ${JSON.stringify(raw)}`);
  }
  return match;
}

function requireObject(value: unknown, label: string): Record<string, unknown> {
  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
    throw new ResponseSchemaError(`${label} is not an object`);
  }
  return value as Record<string, unknown>;
}

function requireObjectField(
  container: Record<string, unknown>,
  key: string,
): Record<string, unknown> {
  return requireObject(container[key], `field '${key}'`);
}

function requireStringField(container: Record<string, unknown>, key: string): string {
  const value = container[key];
  if (typeof value !== 'string') {
    throw new ResponseSchemaError(`field '${key}' is missing or not a string`);
  }
  return value;
}

function requireBooleanField(container: Record<string, unknown>, key: string): boolean {
  const value = container[key];
  if (typeof value !== 'boolean') {
    throw new ResponseSchemaError(`field '${key}' is missing or not a boolean`);
  }
  return value;
}

function requireStringListField(
  container: Record<string, unknown>,
  key: string,
): readonly string[] {
  const value = container[key];
  if (value === undefined) {
    return [];
  }
  if (!Array.isArray(value) || !value.every(item => typeof item === 'string')) {
    throw new ResponseSchemaError(`field '${key}' is not a list of strings`);
  }
  return value;
}

function optionalString(container: Record<string, unknown>, key: string): string {
  const value = container[key];
  return typeof value === 'string' ? value : '';
}
